src/Features/LanguageServer/Protocol/LanguageServerProtocol.cs

```diff
@@ -0,0 +1,224 @@
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer
+{
+    /// <summary>
+    /// Implements Language Server Protocol
+    /// TODO - Make this public when we're ready.
+    /// </summary>
+    [Shared]
+    [Export(typeof(LanguageServerProtocol))]
+    internal sealed class LanguageServerProtocol
+    {
+        private readonly ImmutableDictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>> _requestHandlers;
+
+        [ImportingConstructor]
+        public LanguageServerProtocol([ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        {
+            _requestHandlers = CreateMethodToHandlerMap(requestHandlers);
+        }
+
+        private static ImmutableDictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>> CreateMethodToHandlerMap(IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        {
+            var requestHandlerDictionary = ImmutableDictionary.CreateBuilder<string, Lazy<IRequestHandler, IRequestHandlerMetadata>>();
+            foreach (var lazyHandler in requestHandlers)
+            {
+                requestHandlerDictionary.Add(lazyHandler.Metadata.MethodName, lazyHandler);
+            }
+
+            return requestHandlerDictionary.ToImmutable();
+        }
+
+        private Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(string methodName, Solution solution, RequestType request,
+            LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            var handler = (IRequestHandler<RequestType, ResponseType>)_requestHandlers[methodName].Value;
+            return handler.HandleRequestAsync(solution, request, clientCapabilities, cancellationToken);
+        }
+
+        /// <summary>
+        /// Answers an implementation request by returning the implementation location(s) of a given symbol.
+        /// https://microsoft.github.io/language-server-protocol/specification#textDocument_implementation
+        /// </summary>
+        /// <param name="solution">the solution containing the request document.</param>
+        /// <param name="request">the request document symbol location.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>the location(s) of the implementations of the symbol.</returns>
+        public Task<object> FindImplementationsAsync(Solution solution, LSP.TextDocumentPositionParams request, LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+            => ExecuteRequestAsync<LSP.TextDocumentPositionParams, object>(LSP.Methods.TextDocumentImplementationName, solution, request, clientCapabilities, cancellationToken);
```

Comment:
>=> [](start = 12, length = 2)

Do we validate the arguments somewhere? E.g. `solution != null`, etc.

Reply:
It's currently already handled only on the liveshare side.  However since these will be exposed eventually, I should probably do some validation here as well.

I'll add that in this PR since it shouldn't change the behavior of the liveshare side.